### DISCUSSION TITLE
Set kb_id with EntityRuler

### DISCUSSION THREAD

**Discussion by kinghuang** [explosion/spaCy#8833] [New Features & Project Ideas]
## Feature description

The EntityRuler component creates entity spans and can set the `label` and `ent_id` attributes of the spans. I would like to be able to also set the `kb_id` attribute. So, the EntityRuler can create rule-based entity spans to go with the trained EntityRecognizer and EntityLInker components.

**Comment by adrianeboyd** [comment]
Are there cases where the `kb_id` is not derivable from the `ent_id`?

**Reply by kinghuang** [reply]
I do actually incorporate `kb_id` information in `ent_id` right now, and then extract and set the `kb_id` from `ent_id` in a custom component. But, it would be preferable if the EntityRuler allowed me to just directly specify and set `kb_id`.

As an aside, it would be nice if the EntityRuler allowed me to set any attribute on the resulting spans, including custom extensions. I stuff a lot of information into the `ent_id` (beyond `kb_id`), and then unpack it from there, because it's the only way to pass additional information about the entity spans.

**Reply by adrianeboyd** [reply]
The `ent_id` is already kind of hacked in API-wise, and I think it might be hairy to extend this further within the entity ruler.

If you want to do this without custom components, you might be able to use `ent_id` + `attribute_ruler` to set any attributes on the matched tokens. You can set token extensions in the attribute ruler, but not span extensions.
